src/hash.cr

```diff
@@ -278,6 +290,35 @@ class Hash(K, V)
     self
   end
 
+  # Returns a new hash consisting of entries for which the block returns true.
+  # ```
+  # h = { "a" => 100, "b" => 200, "c" => 300 }
+  # h.select {|k,v| k > "a"}  #=> {"b" => 200, "c" => 300}
+  # h.select {|k,v| v < 200}  #=> {"a" => 100}
+  # ```
+  def select(&block : K, V -> U)
+    each_with_object({} of K => V) do |memo, k, v|
+      memo[k] = v if yield k, v
+    end
+  end
+
+  # Equivalent to `Hash#select` but makes modification on the current object rather that returning a new one.
+  def select!(&block : K, V -> U)
+    delete_if{ |k, v| !yield k, v }
+  end
+
+  # Same as `Hash#delete_if`, but works on (and returns) a copy of the `hsh`. Equivalent to `hsh.dup.delete_if`.
+  def reject(&block : K, V -> U)
+    each_with_object({} of K => V) do |memo, k, v|
+      memo[k] = v unless yield k, v
+    end
+  end
+
+  # Equivalent to `Hash#delete_if`, but returns nil if no changes were made.
```

Comment:
this specific syntax, no, but i have been known to do something like
```rb
unchanged = foo.reject!( ... )
do_something unless unchanged
```

which can also be written like
```rb
new_foo = foo.reject( ... )
unchanged = new_foo.size == foo_size
do_something unless unchanged
```

which is arguably cleaner, albeit less performant.